grails-app/services/org/pih/warehouse/product/ProductService.groovy

```diff
@@ -1228,15 +1228,40 @@ class ProductService {
         return productInstance
     }
 
+    Attribute findExternalProductIdAttribute(String externalAttributeCode = null){
+        if (!externalAttributeCode) {
+            externalAttributeCode = grailsApplication.config.externalProductId.attribute.code
+        }
+        if (!externalAttributeCode) {
+            externalAttributeCode = "EXTERNAL_PRODUCT_ID"
+        }
+        Attribute attribute = Attribute.findByCode(externalAttributeCode)
+        return attribute
+    }
+
+    Product findProductByExternalId(String externalProductId){
+        String externalAttributeCode = grailsApplication.config.externalProductId.attribute.code
+        if(!externalAttributeCode){
+            externalAttributeCode = "EXTERNAL_PRODUCT_ID"
+        }
+        Attribute attribute = findExternalProductIdAttribute(externalAttributeCode)
+        if(!attribute){
+            attribute = new Attribute(code: externalAttributeCode, name: externalAttributeCode, allowOther: false)
+            attribute.save(flush: true)
+        }
+        ProductAttribute productAttribute = ProductAttribute.findByAttributeAndValue(attribute, externalProductId)
```

Comment:
This is fine for now, but what do you think we should do if we find more than one?

Reply:
this will return only one result. so if we have more than that is error, we may did something wrong. that why external id will be duplicate.